### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by SferaDev**
### Motivation

This is just some code I had lying around locally that might be useful or not, also some notes from what I tested.

I initially began reading ```app-runtime```'s code and how they did everything. After understanding what they have accomplished I created my own implementation on-top of ```cross-fetch``` (```whatwg-fetch``` and ```node-fetch```). But I found some blocking issues that appear to break compatibility of features we want on node.js. Finally tried an approach on-top of ``axios`` .

### Lessons learnt

**Cancellation of requests**

```fetch``` (and ```app-runtime```) accomplishes this with a Signal attached on the request. Once the signal (function) is called, the request is aborted. The main problem with this signal is that is an instance of ```AbortController``` which is not included in node. Even though there are some polyfills I couldn't manage to get it working as calling the signal without a window object always returned ```Illegal invocation``` error and crashed the whole test application. I imagine we can make this work somehow if we want to stick with fetch.

**Query string parameters**

``app-runtime`` has several utility methods to translate the JS objects and arrays into query string. I began with them but I did not like how they were implemented. One of the biggest problems of ```fetch``` is that you need to create the URL with query parameters before sending the request. On my last attempt I found out that ```axios``` directly uses ``qs``  to parse them and is more user friendly.

**Caching**

There are several libraries that offer almost-zero configuration for caching. But I would suggest to build our own LRU cache with key mapping (url + serialized query string). Also I would expose in the api an option to the calls ``useCache`` to disable it on-demand per-request. If we finally choose axios this can be easily implemented with an interceptor.

**Automatic retries**

It is a known issue that DHIS2 returns errors 500 randomly, specially if doing demanding requests or during an on-going maintenance task like analytics. I would add a count-based approach with exponential back-off time on a ```catch``` clause. Finally if we reach the limit of retries (5 for example) we trow an error that can be catch'd before returning to the user. If we finally choose axios this can be easily implemented with an interceptor.

**API exposure**

The code I've uploaded exposes a CRUD of an "engine" that includes what we need (credential management, caching, retrying, cancellability/...). It not only offers the aliases for ``get``, ``post``, etc operations but also aims to expose advanced often-used methods for ``import`` and ``export`` metadata. Also we could expose some of the most used properties inside the schemas, systemInfo and me endpoints as utility methods.

**React implementation and hooks**

I really liked the approach of ``app-runtime`` for this, I believe we can follow it on our implementation. On react they create a ``context`` and initialize the API with a <Provider /> on the root of the app similarly to what we do with ReactRouter or the Providers for the Snackbar/Loading. Afterwards the hook they have calls the api from within the react context they did create. The hook is defined in https://github.com/dhis2/app-runtime/blob/master/services/data/src/react/hooks/useQueryExecutor.ts for reference.

**Data store**

I would suggest to create our own implementation of a dataStore ```store``` (pun intended) that mimicks the same collection/element relationship as any other metadata type inside DHIS2. Meaning that the d2-api could handle the creation and maintenance of the different files inside the dataStore namespace instead of leaving all the configuration in the application side. We already discussed the viability of this in the past but we should consider review it when it is used in newer applications. Advanced topics like pagination are the problem but I don't believe it is necessary.

**Sharing settings**

For metadata-sync I'm now developing a sharing settings middleware to both show the visual component and do the update logic. I would love to extract that kind of features and migrate it to d2-api as I find it would be useful in the future.

### CODE REVIEW

File: src/api/d2-api-models.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        const { cancel, response } = this.d2Api.get<any>(`/${this.name}`, params);
+        const responseWithObjects = response.then(response_ => ({
+            ...response_,
+            data: {
+                pager: response_.data.pager,
+                objects: response_.data[this.name],
+            },
+        })) as Promise<AxiosResponse<PaginatedObjects<T>>>;
+        return { cancel, response: responseWithObjects };
+    }
```
Comment (SferaDev) [inline src/api/d2-api-models.ts]: [proposal]

```ts
get<T>(options: GetOptionValue): D2ApiResponse<PaginatedObjects<T>> {
        const params = getParams(options);
        const { cancel, response: rawResponse } = this.d2Api.get<any>(`/${this.name}`, params);
        const response = rawResponse.then(({data, ...rest}) => ({
            ...rest,
            data: {
                pager: data.pager,
                objects: data[this.name],
            },
        })) as Promise<AxiosResponse<PaginatedObjects<T>>>;
        return { cancel, response };
    }
```

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Default export of D2Api
+import D2Api from "./src/api/d2-api";
```
Comment (SferaDev) [inline index.ts]: Not sure about the file names:

I find the d2-api-metadata and d2-api-models to be reduntant. I guess metadata and models would suffice.

File: src/api/d2-api-metadata.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+export default class D2ApiMetadata {
+    d2Api: D2Api;
+
+    constructor(d2Api: D2Api) {
+        this.d2Api = d2Api;
+    }
+
+    get(options: GetOptions) {
+        const metadataOptions = _(options)
+            .map((modelOptions, modelName) => getParams(modelOptions, modelName))
```
Comment (SferaDev) [inline src/api/d2-api-metadata.ts]: Would you do the ``chunk`` for all ``in`` operators in the d2Api.request or rather only here for metadata get?

File: index.ts (hunk 2 above)
Comment (tokland) [reply index.ts]: Good point, renamed.

File: src/api/d2-api-metadata.ts (hunk 3 above)
Comment (tokland) [reply src/api/d2-api-metadata.ts]: I'l take a note on that and work on this extra feature incrementally. The idea is that d2Api remains agnostic about params, so it should be taken care of in the model and metadata classes.

**Comment by tokland** [comment]
Work done:

- [ ] Added extra types to the schemas for COMPLEX types.
- [ ] Sections: api.metadata (/api/metadata, supports GET/POST), api.models (/api/someModelHere, supports GET/POST/PUT/DELETE).

Type inteference will be implemented in another PR.

@adrianq  Ready for review.

File: src/api/d2-api-models.ts (hunk 1 above)
Comment (tokland) [reply src/api/d2-api-models.ts]: I hadn't see that. I refactored it to a `mapD2ApiResponse` function (in FP we would create a Functor :))

**Pull request merged**